prototype/distributor/src/ServerHandler.java

```diff
@@ -55,188 +78,168 @@ public boolean onHandshake(Control.Handshake handshake_packet) {
 
-        General.FvPacket replyOuter =
-            General.FvPacket.newBuilder()
-                .setType(General.FvPacketType.HEARTBEAT)
-                .setInnerPacket(ByteString.copyFrom(heartbeat.toByteArray()))
-                .build();
-
-        sendProto(replyOuter);
+        sendProto(ByteString.copyFrom(heartbeat.toByteArray()), General.FvPacketType.HEARTBEAT);
     }
 
     public void onSessionRequest(Control.SessionRequest sr) {
         Control.ServerMessage reply;
-        General.FvPacket replyOuter;
 
         System.out.println("Session Request packet");
 
         // if user not found
         if (!UsersData.keySet().contains(sr.getName())) {
             reply = Control.ServerMessage.newBuilder()
-                        .setType(Control.ServerMessageType.USER_NOT_FOUND)
+                    .setType(Control.ServerMessageType.USER_NOT_FOUND)
+                    .setSuccess(false)
+                    .build();
+            sendProto(ByteString.copyFrom(reply.toByteArray()), General.FvPacketType.SERVER_MESSAGE);
+            return;
+        }
+
+        //If the user is already in a session
+        for (SessionData session : UsersData.getSessionList()){
+             if (session.getClientList().contains(sr.getName())) {
+                reply = Control.ServerMessage.newBuilder()
+                        .setType(Control.ServerMessageType.USER_ALREADY_IN_SESSION)
                         .setSuccess(false)
                         .build();
+                sendProto(ByteString.copyFrom(reply.toByteArray()), General.FvPacketType.SERVER_MESSAGE);
+
+                return;
+            }
+        }
+
+        //Sends a SESSION_REQUEST packet to the host containing the name of the user trying to connect
+        Control.SessionRequest.Builder srbuilder = Control.SessionRequest.newBuilder();
+        srbuilder.setName(selfName);
+        Control.SessionRequest newRequest = srbuilder.build();
+        sendProto(newRequest.toByteString(), General.FvPacketType.SESSION_REQUEST, UsersData.getUser(sr.getName()));
+    }
 
-            // set outer packet
-            replyOuter = General.FvPacket.newBuilder()
-                             .setType(General.FvPacketType.SERVER_MESSAGE)
-                             .setInnerPacket(ByteString.copyFrom(reply.toByteArray()))
-                             .build();
+    public void onSessionResponse(Control.SessionResponse sr) {
+        Control.ServerMessage reply;
 
-            sendProto(replyOuter);
+        System.out.println("Session Response packet");
 
+        //Connection refused by host
+        if (!sr.getResponse()){
+            reply = Control.ServerMessage.newBuilder()
+                    .setType(Control.ServerMessageType.SESSION_REJECTED)
+                    .setSuccess(false)
+                    .build();
+            sendProto(ByteString.copyFrom(reply.toByteArray()), General.FvPacketType.SERVER_MESSAGE);
```

Comment:
This is sending the "rejected" message back to the prospective host, not the person who requested.

Reply:
Resolved